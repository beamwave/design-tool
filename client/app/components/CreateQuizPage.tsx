import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import fontawesome from '@fortawesome/fontawesome'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { startCreateNewQuiz } from '../actions/create'

interface IProps {
  history?: any
  username?: any
  quizzes?: any
  startCreateNewQuiz: (quiz: any, any) => any
}

export class CreateQuizPage extends Component<IProps, any> {
  state = {
    quiz: {
      author: this.props.username,
      title: '',
      description: '',
      isPrivate: true,
      type: 'quiz'
    },
    quizzes: this.props.quizzes
  }

  private updateQuiz = (e: any, arg1: string) => {
    switch (arg1) {
      case 'title':
        this.setState({
          quiz: {
            ...this.state.quiz,
            title: e.target.value
          }
        })
        break
      case 'description':
        this.setState({
          quiz: {
            ...this.state.quiz,
            description: e.target.value
          }
        })
      default:
        break
    }
    // console.log(this.state)
  }

  private updatePrivate = (e: any) => {
    switch (e.target.value) {
      case 'private':
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: true
          }
        })
        break

      default:
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: false
          }
        })
        break
    }
  }

  private submitQuiz = async (e: any) => {
    e.preventDefault()
    const { startCreateNewQuiz, username } = this.props
    const { quiz } = this.state

    // save quiz to db and update quizzes
    const createAction = await startCreateNewQuiz(quiz, this.props.history)

    // this.props.history.push('/quizzes/created')
    // this.props.history.push('/edit-quiz/' + uuid)
    // .then(res => {
    //   console.log('add quiz props')
    //   console.log(this.props)
    //   this.props.history.push('/add-question')
    // })
    // .catch(err =>
    //   console.log(
    //     'make sure the user knows that the quiz did not make it through'
    //   )
    // )
  }

  // @ts-ignore
  render = () => {
    const {
      quiz: { title }
    } = this.state
    return (
      <div className="create-quiz-page">
        <main>
          <div
            className={
              title.length > 0
                ? 'create-quiz-container valid'
                : 'create-quiz-container invalid'
            }
          >
            <p className="title">Create a quiz</p>
            <form className="details">
              <div className="group">
                <label htmlFor="create-quiz-title">Title</label>
                <input
                  type="text"
                  id="create-quiz-title"
                  value={this.state.quiz.title}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'title')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="create-quiz-description">Description</label>
                <textarea
                  id="create-quiz-description"
                  value={this.state.quiz.description}
                  rows={3}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'description')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="owner">Owner</label>
                <FontAwesomeIcon className="icon" icon="angle-down" />
                <select
                  name="owner"
                  id="question-options-dropdown"
                  onChange={this.updatePrivate}
                >
                  <option value="choose" hidden>
                    select
                  </option>
                  <option value="private">private</option>
                  <option value="public">public</option>
                </select>
              </div>
              <button
                type="button"
                className="save-quiz"
                onClick={this.submitQuiz}
                disabled={title.length > 0 ? false : true}
              >
                Create
              </button>
            </form>
          </div>
        </main>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  username: state.auth.username,
  quizzes: state.quizzes.all
})

export default connect<any, IProps>(
  mapStateToProps,
  { startCreateNewQuiz }
)(CreateQuizPage)
