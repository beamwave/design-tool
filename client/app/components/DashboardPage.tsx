import React, { Component } from 'react'
import { connect } from 'react-redux'
import ReactGridGallery from 'react-grid-gallery'
import { FontAwesomeIcon } from '../../../node_modules/@fortawesome/react-fontawesome'
import { check } from '../helpers/helpers'
import { filter } from '../actions/filter'

interface FData {
  category: string
  value: string
}

interface IProps {
  username: string
  category: string
  categories?: any
  images: any[]
  filters?: any[]
  filter?: (by: FData) => any
  history?: any
}

interface IState {
  images: any[]
}

export class DashboardPage extends Component<IProps, IState> {
  state = {
    images: [],
    categories: 'map an object'
  }

  // filter images
  handleFilterChange = ({ target }) => {
    const { filters, filter } = this.props
    // "medium", "device", "wardrobe", "mood", etc.
    // console.log('dataset', target.dataset.filterName)
    // console.log('value', target.value)

    const by = {
      category: target.dataset.filterName,
      value: target.value
    }

    filter(by)
  }

  handleThumbnailClick = (i, e) => {
    const { images, history } = this.props
    const imageId = images[i]._id
    history.push(`/image/${imageId}`)
  }

  componentDidMount = () => {
    const { images: imagesFromServer } = this.props
    console.log('imagesFromServer', imagesFromServer)
    let images
    if (check(imagesFromServer) && Array.isArray(imagesFromServer))
      images = imagesFromServer.map(image => ({
        src: image.url,
        thumbnail: image.url,
        thumbnailHeight: image.height,
        thumbnailWidth: image.width
      }))

    this.setState({ images })
  }

  render = () => {
    const {
      category,
      categories,
      images: imagesFromServer,
      filters
    } = this.props
    const { images } = this.state
    const alive = check(imagesFromServer)
    return (
      <main className="dashboard-page">
        {
          <div>
            <section className="image-results">
              {alive &&
                imagesFromServer.length === 0 && (
                  <p className="null-case">No images to display.</p>
                )}
              {alive &&
                imagesFromServer.length > 0 && (
                  <ReactGridGallery
                    images={check(imagesFromServer) ? images : []}
                    enableImageSelection={false}
                    onClickThumbnail={this.handleThumbnailClick}
                  />
                )}
            </section>
            <aside className="primary-config">
              {category === 'art' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">medium</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['medium']}
                      data-filter-name="medium"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('medium') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">device</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['device']}
                      data-filter-name="device"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('device') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">project</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['project']}
                      data-filter-name="project"
                      onChange={this.handleFilterChange}
                    >
                      {categories !== undefined &&
                      categories.hasOwnProperty('device') ? (
                        categories['medium'].map(option => (
                          <option key={option.name} value={option.name}>
                            {option.name}
                          </option>
                        ))
                      ) : (
                        <option value="n/a">n/a</option>
                      )}
                    </select>
                  </form>
                </main>
              )}
              {category === 'comments' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">topic</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['topic']}
                      data-filter-name="topic"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">style</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['style']}
                      data-filter-name="style"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">site</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['site']}
                      data-filter-name="site"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
              {category === 'clothes' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">wardrobe</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['wardrobe']}
                      data-filter-name="wardrobe"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">attire</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['attire']}
                      data-filter-name="attire"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
              {category === 'people' && (
                <main>
                  <form className="filter-container">
                    <label htmlFor="select">mood</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['mood']}
                      data-filter-name="mood"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">group</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['group']}
                      data-filter-name="group"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                  <form className="filter-container">
                    <label htmlFor="select">angle</label>
                    <div className="angle-down">
                      <FontAwesomeIcon icon="angle-down" className="icon" />
                    </div>
                    <select
                      name="select"
                      className="select"
                      value={filters['angle']}
                      data-filter-name="angle"
                      onChange={this.handleFilterChange}
                    >
                      <option value="art">art</option>
                      <option value="comments">comments</option>
                      <option value="clothes">clothes</option>
                      <option value="people">people</option>
                    </select>
                  </form>
                </main>
              )}
            </aside>
            <aside className="secondary-config">
              {category === 'art' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="color">color</label>
                      <input
                        className="input"
                        name="color"
                        type="text"
                        placeholder="#hex or name"
                      />
                      <div className="color-group">
                        <div className="color-1" />
                        <div className="color-2" />
                        <div className="color-3" />
                        <FontAwesomeIcon icon="angle-right" className="icon" />
                      </div>
                    </div>
                  </div>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="length">length</label>
                      <input name="length" type="text" placeholder="px" />
                    </div>
                    <div className="input-group">
                      <label htmlFor="width">width</label>
                      <input name="width" type="text" placeholder="px" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'comments' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="username">username</label>
                      <input className="names" name="username" type="text" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'clothes' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="color">color</label>
                      <input
                        className="input"
                        name="color"
                        type="text"
                        placeholder="#hex or name"
                      />
                      <div className="color-group">
                        <div className="color-1" />
                        <div className="color-2" />
                        <div className="color-3" />
                        <FontAwesomeIcon icon="angle-right" className="icon" />
                      </div>
                    </div>
                  </div>
                  <div className="filter-group gender">
                    <div className="input-group">
                      <label>gender</label>
                      <FontAwesomeIcon icon="male" className="male-icon" />
                    </div>
                    <div className="input-group">
                      <FontAwesomeIcon icon="female" className="female-icon" />
                    </div>
                  </div>
                </main>
              )}
              {category === 'people' && (
                <main>
                  <div className="filter-group">
                    <div className="input-group">
                      <label htmlFor="name">name</label>
                      <input className="names" name="name" type="text" />
                    </div>
                  </div>
                  <div className="filter-group gender">
                    <div className="input-group">
                      <label>gender</label>
                      <FontAwesomeIcon icon="male" className="male-icon" />
                    </div>
                    <div className="input-group">
                      <FontAwesomeIcon icon="female" className="female-icon" />
                    </div>
                  </div>
                </main>
              )}
            </aside>
          </div>
        }
      </main>
    )
  }
}

const mapStateToProps = (state, props) => ({
  token: state.auth.token,
  username: state.auth.username,
  category: state.app.category,
  categories: state.images.categories,
  images: state.images.all,
  filters: state.filters
})

export default connect<any>(
  mapStateToProps,
  { filter }
)(DashboardPage)
