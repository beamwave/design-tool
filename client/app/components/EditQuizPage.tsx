import React, { Component } from 'react'
import moment from 'moment'
import numeral from 'numeral'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import Spinner from 'react-spinkit'
import Dropzone from 'react-dropzone'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { startDeleteQuestion, startEditQuestion } from '../actions/questions'
import { editStoreQuiz, startUpdateQuestionsDisplay } from '../actions/quizzes'
import AddQuestion from './AddQuestion'
import { loadModal } from '../actions/modal'
import { DELETE_QUIZ_MODAL } from '../constants/modaltypes'

interface IProps {
  history?: any
  username: any
  quiz: any
  quizzes: any[]
  editStoreQuiz: (any) => any
  adding?: (any) => any
  editFunc?: (any) => any
  startEditQuestion: (any) => any
  startDeleteQuestion: (author: string, title: any) => any
  enableSaveAllButton?: (any) => any
  loadModal?: (type: string, title: string, uuid: any) => any
}

const colors = [
  '#f26161',
  '#cc5252',
  '#e573b1',
  '#ff9d66',
  '#2ddbbc',
  '#42a6a6',
  '#2d9cdb',
  '#3f388c',
  '#4775b2',
  '#b866cc',
  '#333ea6',
  '#f2c94c'
]

export class EditQuizPage extends Component<IProps> {
  state = {
    errMsg: '',
    displayPhoto: '',
    page: 1,
    clickedQuestion: {
      author: '',
      title: '',
      uuid: '',
      format: '',
      answers: [
        {
          answer: '',
          percentPoints: 0,
          feedback: ''
        }
      ]
    },
    questionNumber: 0,
    updatedQuestions: [],
    clickedAddQuestion: false,
    edited: false,
    answerQue: ''
  }

  params = window.location.href.split('/')
  quizUUID = this.params[4]

  page1 = () => this.setState({ page: 1 })

  page2 = () => this.setState({ page: 2 })

  deleteQuizModal = title => (e: any) => {
    this.props.loadModal(DELETE_QUIZ_MODAL, title, null)
    this.props.history.push('/quizzes/created')
  }

  private updateArr = (e: any, arg1: number, arg2: string) => {
    let newAnswersArr = this.state.clickedQuestion.answers
    newAnswersArr[arg1][arg2] = e.target.value
    this.setState({
      edited: true,
      currentQuestion: {
        ...this.state.clickedQuestion,
        answers: newAnswersArr
      }
    })
  }

  public editFunc = () => this.setState({ edited: true })
  public adding = () => this.setState({ clickedAddQuestion: false })

  // public enableSaveAllButton = answerQue => this.setState({ answerQue })

  private updateQuiz = async (e: any) => {
    let sendQuestionList = []
    for (let item of this.state.updatedQuestions) {
      if (item.tags) {
        let set = new Set(
          item.tags.split(',').map(string => {
            return string.trim()
          })
        )

        let testArr: any[] = [...set]
        sendQuestionList.push({
          ...item,
          tags: testArr
        })
      } else {
        sendQuestionList.push({
          ...item
        })
      }
    }
    for (let item of sendQuestionList) {
      console.log('here')
      const data = {
        quizID: window.location.href.split('/')[4],
        question: item
      }

      let testvar = await this.props.startEditQuestion(data)
      console.log(testvar['response']['status'])

      let errorHandling = () => {
        if (testvar['response']['status'] === 400) {
          this.setState({
            ...this.state,
            errMsg:
              'Please make sure all boxes have been filled in and re-submit.'
          })
        } else if (testvar['response']['status'] === 502) {
          this.setState({
            ...this.state,
            errMsg:
              'Your selected image is too large. Please upload a smaller image.'
          })
        } else if (testvar['response']['status'] === 200) {
          this.props.history.push('/quizzes/created')
        }
      }

      errorHandling()
    }
  }

  private saveChangeToState = (e: any) => {
    let newQArr = this.state.updatedQuestions
    newQArr.push(this.state.clickedQuestion)
    this.setState({
      ...this.state,
      updatedQuestions: newQArr,
      errMsg: '',
      edited: false
    })
    this.page1()
    // this.updateStore(this.state.clickedQuestion);
  }

  private deleteQuestion = (e: any) => {
    this.deleteFromStore()
    this.props.startDeleteQuestion(
      this.state.clickedQuestion.author,
      this.state.clickedQuestion.title
    )
  }

  private deleteFromStore = () => {
    let modQuiz = this.props.quiz
    for (let i = 0; i < modQuiz.questions.length; i++) {
      if (modQuiz.questions[i].uuid === this.state.clickedQuestion.uuid) {
        modQuiz.questions.splice(i, 1)
        //here is where we delete a question
      }
    }
    let quizList = this.props.quizzes
    for (let i = 0; i < quizList.length; i++) {
      if (quizList[i].uuid === this.props.quiz.uuid) {
        quizList.splice(i, 1, modQuiz)
      }
    }
    this.props.editStoreQuiz(quizList)
  }

  private updateStore = clickedQuestion => {
    if (this.props.quizzes !== [] && this.props.quiz) {
      let modQuiz = this.props.quiz
      for (let i = 0; i < modQuiz.questions.length; i++) {
        if (modQuiz.questions[i].uuid === this.state.clickedQuestion.uuid) {
          modQuiz.questions = modQuiz.questions.concat(clickedQuestion)
        }
      }

      let quizList = this.props.quizzes
      for (let i = 0; i < quizList.length; i++) {
        if (quizList[i].uuid === this.props.quiz.uuid) {
          quizList.splice(i, 1, modQuiz)
        }
      }
      this.props.editStoreQuiz(quizList)
    }
  }

  public showQuizQuestion = (question: any, count: number, e: any) => {
    e.preventDefault()

    this.setState({
      ...this.state,
      questionNumber: count,
      clickedQuestion: question,
      clickedAddQuestion: false,
      displayPhoto: question.image || ''
    })
  }

  private setAddQuestion = (e: any) => {
    this.setState({
      // we may not need this because if the component reloads this will be cleared anyway
      ...this.state,
      clickedQuestion: {
        author: '',
        title: '',
        uuid: '',
        format: '',
        answers: [
          {
            answer: '',
            percentPoints: 0,
            feedback: ''
          }
        ]
      },
      clickedAddQuestion: true
    })
  }

  private photoUpload: HTMLInputElement

  fileSelectedHandler = e => this.setState({ selectedFile: e.target.files[0] })

  private updateTags = (e: any) => {
    const tags = e.target.value
    this.setState({
      clickedQuestion: {
        ...this.state.clickedQuestion,
        tags: tags
      }
    })
  }

  private onDrop = (files: any) => {
    const file = files[0]
    const getBase64 = async file => {
      var reader = new FileReader()
      reader.readAsDataURL(file)
      let codeString
      reader.onload = function() {
        codeString = reader.result
        codeString = codeString.split(',', 2)
        this.setState({
          ...this.state,
          clickedQuestion: {
            ...this.state.clickedQuestion,
            image: codeString[1],
            newImage: 'yes'
          },
          displayPhoto: codeString.join(',')
        })
      }.bind(this)
      reader.onerror = function(error) {
        console.log('Error: ', error)
      }
      return codeString
    }
    getBase64(file)
  }

  // @ts-ignore
  render = () => {
    const { username, quizzes } = this.props
    let quiz

    if (quizzes) {
      for (let testQuiz of quizzes) {
        if (
          testQuiz !== null &&
          testQuiz.uuid === window.location.href.split('/')[4]
        ) {
          console.log('inside quiz assignment')
          quiz = testQuiz
        }
      }
    }
    console.log('this is the quiz being passed', quiz)
    const {
      page,
      clickedQuestion,
      questionNumber,
      clickedAddQuestion,
      displayPhoto
    } = this.state
    let count = 0
    return (
      <div className="edit-quiz-page">
        {(quiz === undefined || quiz.tags === undefined) && (
          <Spinner className="loading-indicator" name="ball-spin-fade-loader" />
        )}
        {quiz !== undefined &&
          quiz.tags !== undefined && (
            <main>
              <div className="quiz-container">
                <h1 className="title">{quiz.title}</h1>
                <div
                  className="close"
                  onClick={this.deleteQuizModal(this.props.quiz.title)}
                >
                  <FontAwesomeIcon icon="trash" />
                  <p className="hint">Permanently delete this quiz</p>
                </div>
                <div className="tags">
                  {quiz.tags.length === 0 && (
                    <p className="tag-null-set">No tags</p>
                  )}
                  {quiz.tags.length > 0 &&
                    quiz.tags.slice(0, 3).map(tag => (
                      <div key={tag.allLowerCase} className="tag">
                        <div
                          className="tag-dot"
                          style={{
                            height: 8,
                            width: 8,
                            borderRadius: 50,
                            backgroundColor:
                              colors[Math.floor(Math.random() * colors.length)]
                          }}
                        />
                        <p className="tag-text">{tag.allLowerCase}</p>
                      </div>
                    ))}
                  {quiz.tags.length > 3 && (
                    <p className="extra-tags">+{quiz.tags.length - 3}</p>
                  )}
                </div>
                {/* <p className="add-tag">+ tag</p> */}
                <div className="questions">
                  {quiz.questions !== undefined &&
                    quiz.questions.map((question, i) => {
                      if (question !== null && question.author === username)
                        return (
                          <div key={'question' + i}>
                            <p
                              className="question"
                              onClick={this.showQuizQuestion.bind(
                                this,
                                quiz.questions[count],
                                count + 1
                              )}
                            >
                              Question {(count += 1)}
                            </p>
                          </div>
                        )
                    })}
                </div>
                <p
                  onClick={e => this.setAddQuestion(e)}
                  className="add-question"
                >
                  + question
                </p>
                {this.state.errMsg && this.state.errMsg}
                <button onClick={this.updateQuiz} className="save-quiz">
                  Save Quiz
                </button>

                {/* <Link
                to={`/take-quiz/${quiz.uuid}`}
                className="unset-anchor nav-link"
              >
                <div style={quizButzest Quiz</div>
              </Link> */}
              </div>

              {clickedQuestion.title && (
                <div key={clickedQuestion.uuid} className="question-container">
                  <p className="title">Edit question</p>
                  <div className="close" onClick={this.deleteQuestion}>
                    <FontAwesomeIcon icon="trash" />
                    <p className="hint">Permanently delete this question</p>
                  </div>
                  {page === 1 && (
                    <form className="details">
                      <div className="group">
                        <label>Question {questionNumber}</label>
                        <p className="question-text">{clickedQuestion.title}</p>
                      </div>
                      <div className="group">
                        <label>Tags</label>
                        <input
                          onChange={this.updateTags}
                          placeholder="Add tags"
                          // placeholder={
                          //   clickedQuestion.tags.length > 0
                          //     ? clickedQuestion.tags.map(
                          //         tag => `${tag.enteredAs}`
                          //       )
                          //     : 'Assign tags to this question'
                          // }
                        />
                      </div>
                      <div
                        className="group photo-container"
                        onClick={() => this.photoUpload.click()}
                        // This is not a thing
                      >
                        <div className="group">
                          <label>Image</label>
                          <Dropzone
                            onDrop={this.onDrop}
                            className={
                              displayPhoto ? 'dropzone' : 'dropzone with-dots'
                            }
                          >
                            <p className={displayPhoto ? 'white-text' : ''}>
                              + Upload
                            </p>
                            {displayPhoto && (
                              <img src={displayPhoto} alt="img" />
                            )}
                          </Dropzone>
                          <input
                            className="file-upload"
                            style={{ display: 'none' }}
                            name="file"
                            type="file"
                            onChange={this.fileSelectedHandler}
                            ref={photoUpload =>
                              (this.photoUpload = photoUpload)
                            }
                          />
                        </div>
                      </div>
                    </form>
                  )}
                  {page === 2 && (
                    <form className="options">
                      <div className="question-fields">
                        {clickedQuestion.answers.map((ans, index) => (
                          <div key={'ans' + index}>
                            <div className="group">
                              <label
                                htmlFor="true-false-answer"
                                className="label"
                              >
                                Choice
                              </label>
                              <textarea
                                id="true-false-answer"
                                value={ans.answer}
                                className="input"
                                onChange={(e: any) => {
                                  this.updateArr(e, index, 'answer')
                                }}
                                data-enable-grammarly="false"
                              />
                            </div>
                            <div className="group">
                              <label
                                htmlFor="true-false-percent-points"
                                className="input"
                              >
                                Score
                              </label>
                              <input
                                id="true-false-percent-points"
                                value={ans.percentPoints}
                                onChange={(e: any) => {
                                  this.updateArr(e, index, 'percentPoints')
                                }}
                                data-enable-grammarly="false"
                              />
                            </div>
                            {/* <div className="group">
                            <label htmlFor="true-false-feed-back">
                              Feedback
                            </label>
                            <textarea
                              id="true-false-feed-back"
                              value={ans.feedback}
                              className="input"
                              onChange={(e: any) => {
                                this.updateArr(e, index, 'feedback')
                              }}
                              data-enable-grammarly="false"
                            />
                          </div> */}
                          </div>
                        ))}
                      </div>
                    </form>
                  )}
                  <div className="page-navigation">
                    <p
                      className={page === 1 ? 'page-1 active' : 'page-1'}
                      onClick={this.page1}
                    >
                      1
                    </p>
                    <p
                      className={page === 2 ? 'page-2 active' : 'page-2'}
                      onClick={this.page2}
                    >
                      2
                    </p>
                  </div>
                  {page === 2 && (
                    <button
                      disabled={this.state.edited ? false : true}
                      onClick={this.saveChangeToState}
                      className="save-question"
                    >
                      Update question
                    </button>
                  )}
                </div>
              )}

              {clickedAddQuestion && (
                <AddQuestion
                  adding={this.adding}
                  edited={this.state.edited}
                  clickedAddQuestion={this.state.clickedAddQuestion}
                  editFunc={this.editFunc}
                  quizID={this.props.quiz.uuid}
                />
              )}
            </main>
          )}
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  username: state.auth.username,
  quiz:
    state.quizzes.all !== undefined &&
    state.quizzes.all.find(
      quiz =>
        (quiz !== null && quiz.uuid === props.match.params.uuid) ||
        (quiz !== null && quiz.uuid === window.location.href.split('/')[4])
    ),
  quizzes: state.quizzes.all
})

export default connect(
  mapStateToProps,
  {
    editStoreQuiz,
    startDeleteQuestion,
    startEditQuestion,
    loadModal
  }
)(EditQuizPage)
