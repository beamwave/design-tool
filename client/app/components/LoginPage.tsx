import React, { Component } from 'react'
import { Link, RouteComponentProps } from 'react-router-dom'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '../../../node_modules/@fortawesome/react-fontawesome'

import { startLogin } from '../actions/auth'

interface ICredentials {
  username: string
  password: string
}

interface IProps {
  history?: any
  startLogin(data: ICredentials): any
}

interface IState {
  username: string
  password: string
  errors: {
    username: string
    password: string
    global?: string
  }
}

export class LoginPage extends Component<IProps, IState> {
  state = {
    username: '',
    password: '',
    errors: {
      username: '',
      password: '',
      global: ''
    }
  }

  onFieldChange = ({ target }) => {
    let { name, value }: { name: keyof IState; value: string } = target
    this.setState({
      [name]: value
    } as any)
  }

  onSubmit = async e => {
    e.preventDefault()

    const { startLogin, history } = this.props

    const initialData = {
      username: this.state.username,
      password: this.state.password
    }

    // check username is not blank
    if (this.state.username.length === 0) {
      this.setState(prevState => ({
        errors: {
          username: 'username field cannot be blank',
          password: prevState.errors.password
        }
      }))
      // ??
    } else if (this.state.username.length > 0) {
      this.setState(prevState => ({
        errors: {
          username: '',
          password: prevState.errors.password
        }
      }))
    }
    // TODO: implement email regex
    // else if (!this.state.email.match(emailRegex)) {
    //   this.setState(prevState => ({
    //     errors: {
    //       email: 'please enter a valid email',
    //       password: prevState.errors.password
    //     }
    //   }))
    // }

    // check password is not blank
    if (this.state.password.length === 0) {
      this.setState(prevState => ({
        errors: {
          username: prevState.errors.username,
          password: 'password field cannot be blank'
        }
      }))
      // ??
    } else if (this.state.password.length > 0) {
      this.setState(prevState => ({
        errors: {
          username: prevState.errors.username,
          password: ''
        }
      }))
    }

    // if checks pass, send user data to the server
    if (this.state.username.length > 0 && this.state.password.length > 0) {
      const { username, password } = this.state
      const arr = [...username]

      const data = {
        username,
        password
      }

      try {
        await startLogin(data)
        console.log('login success')
        history.push('/dashboard')
      } catch (err) {
        console.log('front end error', err)
        return this.setState({ errors: err.response.data.errors })
      }
    }
  }

  // @ts-ignore
  render = () => {
    const { errors } = this.state
    return (
      <div className="login-page">
        <div className="login-bg" />
        <Link to="/" className="header">
          <div className="globe">
            <FontAwesomeIcon icon="globe" />
          </div>
          <h1 className="app-name">Arsenal</h1>
        </Link>
        <form
          className="login-form"
          onChange={this.onFieldChange}
          onSubmit={this.onSubmit}
        >
          {errors.global && <p className="global-errors">{errors.global}</p>}
          <div className="input-group">
            <label className="title">username</label>
            <input
              type="text"
              name="username"
              style={{
                border: !!errors.username ? '2px solid #e87c7c' : 'none'
              }}
            />
            {errors.username && (
              <p className="inline-errors">{errors.username}</p>
            )}
          </div>
          <div className="input-group">
            <label className="title">password</label>
            <input
              type="password"
              name="password"
              style={{
                border: !!errors.password ? '2px solid #e87c7c' : 'none'
              }}
            />
            {errors.password && (
              <p className="inline-errors">{errors.password}</p>
            )}
          </div>
          <div className="input-group">
            <div className="switch-page">
              <p className="text">Don't have an account?</p>
              <Link to="/signup" className="signup-link">
                &nbsp;Signup
              </Link>
            </div>
            <button type="submit" className="login-button button">
              Login
            </button>
            {/* <GoogleButton onClick={() => {console.log('Google Button clicked!')}} /> */}

            {/* <GoogleLogin 
              clientId='636482594885-49889hn7ee5or2eduu5lrkq5ncngflng.apps.googleusercontent.com'
              onSuccess={responseGoogle}
              onFailure={responseGoogle}
             /> */}
          </div>
        </form>
        <div className="overlay" />
        <div className="bg" />
      </div>
    )
  }
}

export default connect(
  null,
  { startLogin }
)(LoginPage)
