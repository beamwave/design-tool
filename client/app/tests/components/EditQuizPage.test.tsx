import React, { Component } from 'react'
import { shallow } from 'enzyme'
import { EditQuizPage } from '../../components/EditQuizPage'

/* add mock data for quiz and quizzes */
/* add mock function for startEditQuestion, startDeleteQuestion, editStoreQuiz */

test('should render EditQuizPage component', () => {
  const wrapper = shallow(
    <EditQuizPage
      username="PinkMonkeyBird"
      quiz="quiz object placeholder"
      quizzes={['quizzes array placeholder']}
      startEditQuestion={() => `function placeholder`}
      startDeleteQuestion={() => `function placeholder`}
      editStoreQuiz={() => `function placeholder`}
    />
  )
  expect(wrapper).toMatchSnapshot()
})
