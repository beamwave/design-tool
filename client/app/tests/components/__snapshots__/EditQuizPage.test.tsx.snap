// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditQuizPage component 1`] = `
<div
  className="edit-quiz-page"
>
  <Spinner
    className="loading-indicator"
    fadeIn="full"
    name="ball-spin-fade-loader"
    noFadeIn={false}
    overrideSpinnerClassName=""
  />
</div>
`;
